### PR TITLE
feat(lidar_centerpoint): add build only option for tensorrt engine

### PR DESCRIPTION
**Pull request by yukke42**
Signed-off-by: yukke42 <yusuke.muramatsu@tier4.jp>

## Description

Add a new feature to build onnx files to tensorrt engine files without inference for CI/CD requirements.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/lidar_centerpoint/README.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ We trained the models using <https://github.com/open-mmlab/mmdetection3d>.
 | `nms_iou_target_class_names`    | list[string] | -             | target classes for IoU-based Non Maximum Suppression          |
 | `nms_iou_search_distance_2d`    | double       | -             | If two objects are farther than the value, NMS isn't applied. |
 | `nms_iou_threshold`             | double       | -             | IoU threshold for the IoU-based Non Maximum Suppression       |
+| `build_only`                    | bool         | `false`       | shutdown node after TensorRT engin file is built              |
```
Comment (wep21) [inline perception/lidar_centerpoint/README.md:46]:
```suggestion
| `build_only`                    | bool         | `false`       | shutdown node after TensorRT engine file is built              |
```

File: perception/lidar_centerpoint/src/node.cpp [hunk 2]
```diff
@@ -121,6 +121,11 @@ LidarCenterPointNode::LidarCenterPointNode(const rclcpp::NodeOptions & node_opti
     stop_watch_ptr_->tic("cyclic_time");
     stop_watch_ptr_->tic("processing_time");
   }
+
+  if (this->declare_parameter("build_only", false)) {
+    RCLCPP_INFO(this->get_logger(), "TensorRT engine is built and shutdown node.");
+    exit(0);
```
Comment (wep21) [inline perception/lidar_centerpoint/src/node.cpp:127]:
```suggestion
    rclcpp::shutdown();
```

File: perception/lidar_centerpoint/launch/lidar_centerpoint.launch.xml [hunk 3]
```diff
@@ -8,6 +8,7 @@
   <arg name="class_remapper_param_path" default="$(find-pkg-share lidar_centerpoint)/config/detection_class_remapper.param.yaml"/>
   <arg name="score_threshold" default="0.35"/>
   <arg name="has_twist" default="false"/>
+  <arg name="build_only" default="false" description=""/>
```
Comment (wep21) [inline perception/lidar_centerpoint/launch/lidar_centerpoint.launch.xml:11]: Is it necessary to add empty description?

Comment (yukke42) [reply perception/lidar_centerpoint/launch/lidar_centerpoint.launch.xml]: added at [a8a7912](https://github.com/autowarefoundation/autoware.universe/pull/2677/commits/a8a791266ab8ccdd98008a8f54ad65f0f9e1128f).

**Comment by yukke42** [comment]
A node cannot be shutdown successfully, I'll investigate this issue.

image

File: perception/lidar_centerpoint/src/node.cpp (hunk 2 above)
Comment (yukke42) [reply perception/lidar_centerpoint/src/node.cpp]: changed at [80be962](https://github.com/autowarefoundation/autoware.universe/pull/2677/commits/80be9622b40409c5f9006ee4104e88b9f0576b1a).

**Comment by knzo25** [comment]
@yukke42 

I think this has to do with the destructor order due to the data relationship of some elements.

If I add this, the problem does away:

```
TensorRTWrapper::~TensorRTWrapper()
{
  context_.reset();
  runtime_.reset();
  plan_.reset();
  engine_.reset();
}
```

**Comment by yukke42** [comment]
> @yukke42
> 
> I think this has to do with the destructor order due to the data relationship of some elements.
> 
> If I add this, the problem goes away:
> 
> ```
> TensorRTWrapper::~TensorRTWrapper()
> {
>   context_.reset();
>   runtime_.reset();
>   plan_.reset();
>   engine_.reset();
> }
> ```
> 
> Log:
> 
> ```
> ros2 launch lidar_centerpoint lidar_centerpoint.launch.xml build_only:=true
> [INFO] [launch]: All log files can be found below /home/kenzolobos/.ros/log/2023-03-13-10-55-11-049175-npc2103019-402737
> [INFO] [launch]: Default logging verbosity is set to INFO
> [INFO] [lidar_centerpoint_node-1]: process started with pid [402745]
> [lidar_centerpoint_node-1] 1678672511.315870 [139] lidar_cent: config: //CycloneDDS/Domain/General: 'NetworkInterfaceAddress': deprecated element (file:///opt/autoware/cyclonedds_config.xml line 5)
> [lidar_centerpoint_node-1] 1678672511.315893 [139] lidar_cent: config: //CycloneDDS/Domain/Internal/MinimumSocketReceiveBufferSize: setting moved to //CycloneDDS/Domain/Internal/SocketReceiveBufferSize[@min] (line 8)
> [lidar_centerpoint_node-1] [INFO] [1678672511.675429330] [tensorrt_common]: [MemUsageChange] Init CUDA: CPU +303, GPU +0, now: CPU 332, GPU 245 (MiB)
> [lidar_centerpoint_node-1] [INFO] [1678672511.705650244] [tensorrt_common]: Loaded engine size: 0 MiB
> [lidar_centerpoint_node-1] [INFO] [1678672511.707445022] [tensorrt_common]: [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> [lidar_centerpoint_node-1] Loaded engine from /home/kenzolobos/workspace/pilot-auto/install/lidar_centerpoint/share/lidar_centerpoint/data/pts_voxel_encoder_centerpoint_tiny.engine
> [lidar_centerpoint_node-1] [INFO] [1678672511.709265274] [tensorrt_common]: [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +157, now: CPU 0, GPU 157 (MiB)
> [lidar_centerpoint_node-1] [INFO] [1678672511.718010348] [lidar_centerpoint]: TensorRT engine is built and shutdown node.
> [INFO] [lidar_centerpoint_node-1]: process has finished cleanly [pid 402745]
> ```
> 
> Additionally, there are some unused variables

@knzo25 Thank you for your advise! I have fixed the error.

File: perception/lidar_centerpoint/README.md [hunk 4]
```diff
@@ -43,6 +43,7 @@ We trained the models using <https://github.com/open-mmlab/mmdetection3d>.
 | `nms_iou_target_class_names`    | list[string] | -             | target classes for IoU-based Non Maximum Suppression          |
 | `nms_iou_search_distance_2d`    | double       | -             | If two objects are farther than the value, NMS isn't applied. |
 | `nms_iou_threshold`             | double       | -             | IoU threshold for the IoU-based Non Maximum Suppression       |
+| `build_only`                    | bool         | `false`       | shutdown node after TensorRT engine file is built             |
```
Comment (knzo25) [inline perception/lidar_centerpoint/README.md:46]:
```suggestion
| `build_only`                    | bool         | `false`       | shutdown the node after TensorRT engine file is built             |
```

**Review by knzo25 — COMMENTED** [review]
Other than a typo, LGTM !
Checked the runtime of the engine generation and posterior loading

**Review by knzo25 — APPROVED** [review]
LGTM!

**Comment by yukke42** [comment]
@wep21 Cloud you check the updates? :bow:

**Review by wep21 — APPROVED** [review]
LGTM

**Pull request merged**